golang: speedup using TRIE

starius opened the pull request:
Most patterns are simple search strings (not special Regexp symbols). Some utilize ^ and $, which can be emulated in plaintext search by appending these characters to the text itself for matching as regular characters. Additionally, some patterns involve (xx|yy) or [xY] structures, which expand to several plaintexts. Rare patterns require real regexp matching.

I've applied these simplifications and modifications. They are detected automatically by function `analyzePattern` which returns the list of plain text patterns (all possible search strings) or the main liternal and a regexp in complex cases. The main literal is needed to know when to run the regexp.

Search strings are substituted with a random hex string of length 16 (to prevent spontaneous or intentional matching with anything), followed by a label ("-" for simple search strings, "*" for rare cases requiring regexp, and a number encoded as "%05d" format).

All replacements are performed using strings.Replacer, which utilizes TRIE and is therefore very fast. The random hex string is searched within the output of the replacement. If it's not found, it indicates a mismatch. If found, it's either a match (for simple search string labels) or a potential match (for regexp patterns). In the latter case, the corresponding regexp is executed on the text to verify the match.

Benchmark comparison:

```
$ benchstat old.txt new.txt
goos: linux
goarch: amd64
pkg: github.com/monperrus/crawler-user-agents
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
						   │   old.stat   │              new.stat               │
						   │    sec/op    │   sec/op     vs base                │
IsCrawlerPositive-2          70.453µ ± 1%   1.508µ ± 1%  -97.86% (p=0.000 n=20)
MatchingCrawlersPositive-2   77.343µ ± 1%   1.585µ ± 1%  -97.95% (p=0.000 n=20)
IsCrawlerNegative-2          75.237µ ± 0%   1.725µ ± 1%  -97.71% (p=0.000 n=20)
MatchingCrawlersNegative-2   75.884µ ± 1%   1.725µ ± 0%  -97.73% (p=0.000 n=20)
geomean                       74.68µ        1.633µ       -97.81%

						   │   old.stat   │                new.stat                 │
						   │     B/s      │      B/s       vs base                  │
IsCrawlerPositive-2          2.141Mi ± 1%   99.955Mi ± 1%  +4568.60% (p=0.000 n=20)
MatchingCrawlersPositive-2   1.950Mi ± 1%   95.067Mi ± 1%  +4774.57% (p=0.000 n=20)
IsCrawlerNegative-2          1.936Mi ± 0%   84.586Mi ± 1%  +4269.21% (p=0.000 n=20)
MatchingCrawlersNegative-2   1.926Mi ± 1%   84.615Mi ± 0%  +4292.33% (p=0.000 n=20)
geomean                      1.987Mi         90.81Mi       +4471.46%
```

New implementation is 40 times faster!

monperrus commented:
thanks @starius 

@javierron are you able to code-review this one?

javierron reviewed (commented):
@monperrus @starius 
Very smart use of Replacer!

The main issue I see is that the list of patterns will need to be updated as more patterns are added to the file, which is not ideal. 
Maybe the solution is to determine at initialization if a pattern is either literal or regex. I believe this would simplify `analizePattern`, and not have too much impact on performance, since most of the patterns are literals anyway.

Other than that, I leave only some general comments.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+	regexps  []regexpPattern
+}
+
+var uniqueToken = hex.EncodeToString((&maphash.Hash{}).Sum(nil))
```
A short string may match spontaneously. A deterministic string may match intentionally: someone attacking the detector can pass a text with that string followed by wrong format in it, causing a panic in the current implementation.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+
+	for {
+		uniquePos := strings.Index(replaced, uniqueToken)
+		if uniquePos == -1 {
```
Can you elaborate on this, please?

```go
for uniquePos := strings.Index(replaced, uniqueToken); uniquePos != -1; uniquePos = strings.Index(replaced, uniqueToken) {
}
```

Like this? I don't really like it, because it copy-pastes the `strings.Index` call.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
 }()
 
 // Returns if User Agent string matches any of crawler patterns.
 func IsCrawler(userAgent string) bool {
```
I thought about it. It would break an API, which is not great. Also I think in this case a panic is in place, because a panic means a mistake in the code [or in the config of an app / system, but not in this case]. Returning an error is for cases of mistake in the input in general (in this case, a mistake in user agent format). The panics that the code has will fire if there is a bug, a mistake in the code.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+	}
+	oldnew = append(oldnew, oldnew2...)
+
+	regexps2 := make([]regexpPattern, len(regexps))
```
Save memory. `regexps` is allocated in `append` calls and it is likely to have extra capacity, because `append` reallocates with reserves in capacity. `regexps2` is allocated for exact size that is needed. We know the size only after finishing iterating over patterns, so we can not preallocate an array of the exact size in advance.

I added a comment in the code.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+			})
+		}
+
+		replaceWith := fmt.Sprintf(" %s%c%05d ", uniqueToken, label, num)
```
Fixed. Moved to const `numLen`.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+		}
+
+		start := uniquePos + len(uniqueToken) + 1
+		if start+5 >= len(replaced) {
```
Fixed. Moved to a const `uniqueTokenLen`.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+			break
+		}
+
+		start := uniquePos + len(uniqueToken) + 1
```
Fixed. Moved to a const `uniqueTokenLen`.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
 }
 
 // Finds all crawlers matching the User Agent and returns the list of their indices in Crawlers.
 func MatchingCrawlers(userAgent string) []int {
```
See my [comment](https://github.com/monperrus/crawler-user-agents/pull/353#discussion_r1563606086) above.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+	for {
+		uniquePos := strings.Index(replaced, uniqueToken)
+		if uniquePos == -1 {
+			break
```
See my [comment](https://github.com/monperrus/crawler-user-agents/pull/353#discussion_r1563601989) above.

starius commented on validate.go:
```diff
@@ -80,31 +84,198 @@ var Crawlers = func() []Crawler {
+		return []string{prefix}, nil
+	}
+
+	mainLiternal, has := pattern2mainLiteral[pattern]
```
Fixed!

starius commented:
Hi @javierron !

Thank you very much for the review!
I addressed the feedback, see my comments.

> Maybe the solution is to determine at initialization if a pattern is either literal or regex. I believe this would simplify analizePattern

Can you elaborate on this, please? `analizePattern` already determines automatically if a pattern is a literal:

```go
prefix, complete := re.LiteralPrefix()
if complete {
  return []string{prefix}, nil
}
```
(I put this code after checking presence in `pattern2literals` table, because it is faster than compiling the regexp, so regexp compilation is not done for cases from `pattern2literals` table.)

We still need both tables `pattern2literals` and `pattern2mainLiteral`, because we need to do something with non-literal regexps. Maybe `pattern2literals` could be generated automatically, but such a code would be quite complicated, so I think it is not worth it, to generate such a small table. And for `pattern2mainLiteral` I don't know how to automate its creation even in theory: we need to find a long enough literal in regexp which is present in any match. It requires in-depth analysis of the structure of regexp. And the hardcoded table is even smaller...

Do you have ideas how to do it in an elegant way?

monperrus commented:
@starius thanks for the updates

@javierron let me know how we should proceed thanks.

starius commented:
Hi @javierron @monperrus !

I think it is possible to implement `func analyzePattern` without hardcoded tables. Go has package [regexp/syntax](https://pkg.go.dev/regexp/syntax) which provides introspection of regular expressions. I think it is possible to use it to convert a pattern to the list of all possible matching strings (what `pattern2literals` does) and to extract longest common sub-string from a regexp (what `pattern2mainLiteral` does). I'll try to implement this in few days and update the PR.

starius commented:
@monperrus Yes. Sorry for the delay :pray:

starius commented:
@monperrus @javierron  @srstsavage Please check the latest version of the PR!
The tables with hardcoded cases were replaced with `analyzePattern` function, which does it automatically.

monperrus commented:
For the record and Github indexing:

Most patterns are simple search strings (not special Regexp symbols).
Some utilize ^ and $, which can be emulated in plaintext search by appending
these characters to the text itself for matching as regular characters.
Additionally, some patterns involve (xx|yy) or [xY] structures, which expand
to several plaintexts. Rare patterns require real regexp matching.

I've applied these simplifications and modifications. They are detected
automatically by function analyzePattern which returns the list of plain text
patterns (all possible search strings) or the main liternal and a regexp in
complex cases. The main literal is needed to know when to run the regexp.

Search strings are substituted with a random hex string of length 16 (to prevent
spontaneous or intentional matching with anything), followed by a label ("-" for
simple search strings, "*" for rare cases requiring regexp, and a number encoded
as "%05d" format).

All replacements are performed using strings.Replacer, which utilizes TRIE and
is therefore very fast. The random hex string is searched within the output of
the replacement. If it's not found, it indicates a mismatch. If found, it's
either a match (for simple search string labels) or a potential match (for
regexp patterns). In the latter case, the corresponding regexp is executed on
the text to verify the match.

Benchmark comparison:

$ benchstat old.txt new.txt
goos: linux
goarch: amd64
pkg: github.com/monperrus/crawler-user-agents
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
                           │   old.stat   │              new.stat               │
                           │    sec/op    │   sec/op     vs base                │
IsCrawlerPositive-2          70.453µ ± 1%   1.508µ ± 1%  -97.86% (p=0.000 n=20)
MatchingCrawlersPositive-2   77.343µ ± 1%   1.585µ ± 1%  -97.95% (p=0.000 n=20)
IsCrawlerNegative-2          75.237µ ± 0%   1.725µ ± 1%  -97.71% (p=0.000 n=20)
MatchingCrawlersNegative-2   75.884µ ± 1%   1.725µ ± 0%  -97.73% (p=0.000 n=20)
geomean                       74.68µ        1.633µ       -97.81%

                           │   old.stat   │                new.stat                 │
                           │     B/s      │      B/s       vs base                  │
IsCrawlerPositive-2          2.141Mi ± 1%   99.955Mi ± 1%  +4568.60% (p=0.000 n=20)
MatchingCrawlersPositive-2   1.950Mi ± 1%   95.067Mi ± 1%  +4774.57% (p=0.000 n=20)
IsCrawlerNegative-2          1.936Mi ± 0%   84.586Mi ± 1%  +4269.21% (p=0.000 n=20)
MatchingCrawlersNegative-2   1.926Mi ± 1%   84.615Mi ± 0%  +4292.33% (p=0.000 n=20)
geomean                      1.987Mi         90.81Mi       +4471.46%

New implementation is 40 times faster!

monperrus commented:
thanks a lot @starius
